src/ows/src/main/java/org/geoserver/ows/Dispatcher.java

```diff
@@ -1530,6 +1530,14 @@ void parseKVP(Request req) throws ServiceException {
     }
 
     Map<String, Object> parseKVP(Request req, Map<String, Object> kvp) {
+        // if the service is not explicitly set, some WFS parsers will not be found
+        if (!kvp.containsKey("service")
+                && !kvp.containsKey("SERVICE")
+                && req.getHttpRequest().getServletPath().toLowerCase().startsWith("/wfs")
```

Comment:
Good start but I'd generalize it a bit, to make it valid across all OGC services.

See what the Dispatcher is already doing [here](https://github.com/geoserver/geoserver/blob/ce22d0d99c89bc265c64e61dda1340ce1ebf63ef/src/ows/src/main/java/org/geoserver/ows/Dispatcher.java#L587) and [here].(https://github.com/geoserver/geoserver/blob/main/src/ows/src/main/java/org/geoserver/ows/Dispatcher.java#L1597). 
It's basically assuming the request path is the service, it if could not be determined in any other way. 
I would do the same, and ignore the path if it happens to be the generic "ows".

This way you have a fix that generalizes for wms, wfs, wcs and so on (take the path, make it uppercase, use as service).

Reply:
Ok, I piggybacked off of the method you pointed me too and expanded it a bit with a check on the HttpRequest if the path is not set (which was the case with KVP parse)